sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/viewer/ViewerApplication.java

```diff
@@ -0,0 +1,341 @@
+    metadata = reader.readMetadata();
+    updateTitle();
+    loadComponents();
+  }
+
+  private void loadComponents() {
+    int rootComponentRef = metadata.getRootComponentRef();
+    Component component = reader.readComponent(rootComponentRef);
+    DefaultMutableTreeNode project = createNode(component);
+    loadChildren(component, project);
+    getComponentTree().setModel(new DefaultTreeModel(project));
+  }
+
+  private DefaultMutableTreeNode createNode(Component component) {
+    DefaultMutableTreeNode node = new DefaultMutableTreeNode(component) {
+      @Override
+      public String toString() {
+        return getNodeName((Component) getUserObject());
+      }
+    };
+    return node;
+  }
+
+  private String getNodeName(Component component) {
+    switch (component.getType()) {
+      case PROJECT:
+      case MODULE:
+        return component.getName();
+      case DIRECTORY:
+      case FILE:
+        return component.getPath();
+      default:
+        throw new IllegalArgumentException("Unknow component type: " + component.getType());
+    }
+  }
+
+  private void loadChildren(Component parentComponent, DefaultMutableTreeNode parentNode) {
+    for (int ref : parentComponent.getChildRefList()) {
+      Component child = reader.readComponent(ref);
+      DefaultMutableTreeNode childNode = createNode(child);
+      parentNode.add(childNode);
+      loadChildren(child, childNode);
+    }
+
+  }
+
+  private void updateTitle() {
+    frame.setTitle(metadata.getProjectKey() + (metadata.hasBranch() ? (" (" + metadata.getBranch() + ")") : "") + " " + sdf.format(new Date(metadata.getAnalysisDate())));
+  }
+
+  private void updateDetails(Component component) {
+    componentEditor.setText(component.toString());
+    updateHighlighting(component);
+    updateSource(component);
+    updateCoverage(component);
+  }
+
+  private void updateCoverage(Component component) {
+    coverageEditor.setText("");
+    if (reader.hasCoverage(component.getRef())) {
+      try (CloseableIterator<BatchReport.Coverage> it = reader.readComponentCoverage(component.getRef())) {
+        while (it.hasNext()) {
+          BatchReport.Coverage coverage = it.next();
+          coverageEditor.getDocument().insertString(coverageEditor.getDocument().getEndPosition().getOffset(), coverage.toString() + "\n", null);
+        }
+      } catch (Exception e) {
+        throw new IllegalStateException("Can't read code coverage for " + getNodeName(component), e);
+      }
+    }
+  }
+
+  private void updateSource(Component component) {
+    File sourceFile = reader.getFileStructure().fileFor(Domain.SOURCE, component.getRef());
+    if (sourceFile.exists()) {
+      try (Scanner s = new Scanner(sourceFile, StandardCharsets.UTF_8.name()).useDelimiter("\\Z")) {
+        sourceEditor.setText(s.next());
+      } catch (IOException ex) {
+        StringWriter errors = new StringWriter();
+        ex.printStackTrace(new PrintWriter(errors));
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Use a logger to log this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1148)